### PR TITLE
Feature temperature rca

### PR DESCRIPTION
**Pull request by hxiao608**
- Description: Enabling Perftop support for Temperature RCA

- How to Run:  
Receive sample data: inside performance-analyzer-rca folder: run `./gradlew runRally`.
1.  Run `./gradlew build -Dbuild.linux={true/false} -Dbuild.macos={true/false}` to build executable
2. Run `build/perf-top-macos --dashboard lib/dashboards/rca/TemperatureAnalysis.json --endpoint localhost:9600 --mode rca`
or:
2. Run `node ./lib/bin.js --dashboard lib/dashboards/rca/TemperatureAnalysis.json --endpoint localhost:9600 --mode rca`

- request url:
`localhost:9600/_opendistro/_performanceanalyzer/rca?name=AllTemperatureDimensions&local=true`

- response:
```
{
    "AllTemperatureDimensions": [
        {
            "NodeLevelDimensionalSummary": [
                {
                    "dimension": "Shard_Size_In_Bytes",
                    "mean": 2,
                    "total": 1415.0,
                    "numShards": 5,
                    "NodeLevelZoneSummary": [
                        {
                            "zone": "HOT",
                            "all_shards": []
                        },
                        {
                            "zone": "WARM",
                            "all_shards": []
                        },
                        {
                            "zone": "LUKE_WARM",
                            "all_shards": [
                                {
                                    "index_name": "pmc",
                                    "shard_id": 0,
                                    "temperature": [
                                        {
                                            "dimension": "Shard_Size_In_Bytes",
                                            "value": "2"
                                        }
                                    ]
                                },
                                {
                                    "index_name": "pmc",
                                    "shard_id": 1,
                                    "temperature": [
                                        {
                                            "dimension": "Shard_Size_In_Bytes",
                                            "value": "2"
                                        }
                                    ]
                                },
                                {
                                    "index_name": "pmc",
                                    "shard_id": 2,
                                    "temperature": [
                                        {
                                            "dimension": "Shard_Size_In_Bytes",
                                            "value": "2"
                                        }
                                    ]
                                },
                                {
                                    "index_name": "pmc",
                                    "shard_id": 3,
                                    "temperature": [
                                        {
                                            "dimension": "Shard_Size_In_Bytes",
                                            "value": "2"
                                        }
                                    ]
                                },
                                {
                                    "index_name": "pmc",
                                    "shard_id": 4,
                                    "temperature": [
                                        {
                                            "dimension": "Shard_Size_In_Bytes",
                                            "value": "2"
                                        }
                                    ]
                                }
                            ]
                        },
                        {
                            "zone": "COLD",
                            "all_shards": []
                        }
                    ],
                    "timestamp": "1594423374666"
                }
            ]
        }
    ]
}
```

- preview
Screen Shot 2020-07-10 at 4 19 33 PM

- Graphs supported (for dimension "Shard_Size_In_Bytes")
1. Donut: to visualize the shard distribution 
2. Line: to reflect the change of mean temperature and total temperature
3. Table: to indicate properties in each node level zone. 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,2 +1,3 @@
 build
 node_modules
+.gradle
```
Comment (sendkb) [inline .gitignore:3]: shouldn't we include .idea files too here

File: .idea/gradle.xml [hunk 2]
```diff
@@ -0,0 +1,21 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (sendkb) [inline .idea/gradle.xml:1]: why are .idea files in this diff

File: lib/bin.js [hunk 3]
```diff
@@ -63,13 +63,22 @@ if (args.dashboard === 'ClusterOverview') {
   jsonData = clusterThreadAnalysisJSON;
 } else if (args.dashboard === 'NodeAnalysis') {
   jsonData = nodeAnalysisJSON;
+} else if(args.dashboard === "TemperatureAnalysis"){
```
Comment (sendkb) [inline lib/bin.js:66]: nit: space after if and ')' and a few other places below. You may wan to fix the styling in your IDE

File: lib/bin.js [hunk 4]
```diff
@@ -63,13 +63,22 @@ if (args.dashboard === 'ClusterOverview') {
...
   jsonData = nodeAnalysisJSON;
+} else if(args.dashboard === "TemperatureAnalysis"){ 
+  jsonData = temperatureAnalysisJSON;
 } else {
   jsonData = require(path.resolve(process.cwd(), args.dashboard));
 }
 
+var dashboard = JSON.stringify(args.dashboard);
+if(dashboard.includes("TemperatureAnalysis")){
+  isTemperatureAnalysis = true;
```
Comment (sendkb) [inline lib/bin.js:74]: why do we need this block. Can we not set isTemperatureAnalysis after line 67 above

File: lib/bin.js [hunk 5]
```diff
@@ -82,4 +91,10 @@ console.error = function (msg) {
   logFile.write(util.format(msg) + '\n');
 };
 
-graphGenerator.initAndStart(jsonData);
\ No newline at end of file
+if(isTemperatureAnalysis){
+  graphGenerator.initAndStartTemperatureAnalysis(jsonData);
```
Comment (sendkb) [inline lib/bin.js:95]: should we create an init method for all RCA outputs instead of hardcoding temp analysis here. Within the RCA output init, we can handle temp analysis. Tomorrow, it can be used to parse jvm or hot node outputs

i would rethink the input args to see if we need a mode flag which can indicate whether to query rca apis or pa metric apis

File: lib/perf-top/util/metric-table.js [hunk 6]
```diff
@@ -133,4 +133,4 @@ function appendMetricUnits (columns, metricUnits) {
   return columns;
 }
 
-module.exports.metricTable = metricTable;
+module.exports.metricTable = metricTable;
```
Comment (sendkb) [inline lib/perf-top/util/metric-table.js:136]: there is no change in this file. remove

File: lib/perf-top/util/validate-query-params.js [hunk 7]
```diff
@@ -52,6 +52,7 @@ function validateTableQueryParams (queryParams) {
   }
 }
 
+
 module.exports.validateTableQueryParams = validateTableQueryParams;
 module.exports.validateBarQueryParams = validateBarQueryParams;
-module.exports.validateLineQueryParams = validateLineQueryParams;
+module.exports.validateLineQueryParams = validateLineQueryParams;
```
Comment (sendkb) [inline lib/perf-top/util/validate-query-params.js:58]: there is no change in this file. remove

File: lib/perf-top/generate-graphs.js [hunk 8]
```diff
@@ -46,10 +50,41 @@ function initAndStart (jsonData) {
         graphs.allGraphs.push(graph);
       }
     }
-    // Generate graph on screen
```
Comment (sendkb) [inline lib/perf-top/generate-graphs.js:49]: revert this change. there is nothing that is modified

File: lib/perf-top/util/metric-bar.js [hunk 9]
```diff
@@ -28,7 +28,6 @@ var queryValidator = require('./validate-query-params.js');
  * @param {object} screen - blessed.screen() object.
  */
 function metricBar (endpoint, gridOptions, queryParams, options, screen) {
-  queryValidator.validateBarQueryParams(queryParams);
```
Comment (sendkb) [inline lib/perf-top/util/metric-bar.js:31]: why is this removed.

File: lib/perf-top/util/metric-bar.js [hunk 10]
```diff
@@ -45,6 +44,7 @@ function metricBar (endpoint, gridOptions, queryParams, options, screen) {
   this.dataTimestamp = {};
 }
 
+
```
Comment (sendkb) [inline lib/perf-top/util/metric-bar.js:47]: no changes in this file. revert

File: lib/perf-top/util/metric-graphs.js [hunk 11]
```diff
@@ -44,12 +42,13 @@ metricGraphs.prototype.resizeGraphsToScreen = function () {
...
 };
 
+
 /**
  * Generate the graphs and update the graphs every refreshInterval milliseconds.
  */
 metricGraphs.prototype.start = function () {
   for (var graph in this.allGraphs) {
-    var metricGraph = this.allGraphs[graph];
+    var metricGraph = this.allGraphs[graph];
```
Comment (sendkb) [inline lib/perf-top/util/metric-graphs.js:51]: no changes here. revert it

File: lib/perf-top/util/metric-line.js [hunk 12]
```diff
@@ -40,8 +40,8 @@ function metricLine (endpoint, gridOptions, queryParams, options, screen) {
 
   this.refreshInterval = (options.refreshInterval > 5000) ? options.refreshInterval : 5000;
 
-  var grid = new contrib.grid({ rows: gridOptions.rows, cols: gridOptions.cols, screen: screen });
-  this.line = grid.set(options.gridPosition.row, options.gridPosition.col, options.gridPosition.rowSpan,
+  var grid = new contrib.grid({ rows: gridOptions.rows, cols: gridOptions.cols, screen: screen }); // /set grid
```
Comment (sendkb) [inline lib/perf-top/util/metric-line.js:43]: no changes here- revert

Comment (hxiao608) [reply lib/perf-top/util/metric-line.js]: Reverted, Thanks

File: lib/perf-top/util/generate-data.js [hunk 13]
```diff
@@ -370,6 +499,45 @@ function removeStaleData(metricData, dataTimestamp) {
   }
 }
 
+/**
```
[the remark this answers is not in the document]
Comment (hxiao608) [reply lib/perf-top/util/generate-data.js:502]: Thank you for the suggestions, I will change the structure.

File: lib/perf-top/rca/util/temperature-profile/generate-data.js [hunk 14]
```diff
@@ -0,0 +1,262 @@
...
+    protocol: parsedURL.protocol,
+    host: parsedURL.hostname,
+    auth: parsedURL.auth,
+    port: (parsedURL.port === null) ? (parsedURL.protocol === 'https:' ? 443 : 80 ): parsedURL.port,
+    path: parsedURL.path
+  };
+}
+
+/**
+ * Makes a HTTP(S) request and returns the response.
```
Comment (aditjind) [inline lib/perf-top/rca/util/temperature-profile/generate-data.js:69]: This code is duplicated across both the modules (RCA & metrics). Lets raise another PR to remove this code duplication.

Comment (hxiao608) [reply lib/perf-top/rca/util/temperature-profile/generate-data.js:69]: Will do!
